pkg/apis/kudo/v1alpha1/instance_types.go

```diff
@@ -16,65 +16,372 @@ limitations under the License.
+	}
 
-// PhaseStateInProgress actively deploying, but not yet healthy.
-const PhaseStateInProgress PhaseState = "IN_PROGRESS"
+	return nil
+}
 
-// PhaseStatePending Not ready to deploy because dependent phases/steps not healthy.
-const PhaseStatePending PhaseState = "PENDING"
+// isUpgradePlan returns true if this could be an upgrade plan - this is just an approximation because deploy plan can be used for both
+func isUpgradePlan(planName string) bool {
+	return planName == "deploy" || planName == "upgrade"
+}
 
-// PhaseStateComplete deployed and healthy.
-const PhaseStateComplete PhaseState = "COMPLETE"
+// UpdateInstanceStatus updates `Status.PlanStatus` and `Status.AggregatedStatus` property based on the given plan
+func (i *Instance) UpdateInstanceStatus(planStatus *PlanStatus) {
+	for k, v := range i.Status.PlanStatus {
+		if v.Name == planStatus.Name {
+			i.Status.PlanStatus[k] = *planStatus
+			i.Status.AggregatedStatus.Status = planStatus.Status
+			if planStatus.Status.IsTerminal() {
+				i.Status.AggregatedStatus.ActivePlanName = ""
+			}
+		}
+	}
+}
 
-// PhaseStateError there was an error deploying the application.
-const PhaseStateError PhaseState = "ERROR"
+const snapshotAnnotation = "kudo.dev/last-applied-instance-state"
 
-// PhaseStateSuspend Spec was triggered to stop this plan execution.
-const PhaseStateSuspend PhaseState = "SUSPEND"
+// SaveSnapshot stores the current spec of Instance into the snapshot annotation
+// this information is used when executing update/upgrade plans, this overrides any snapshot that existed before
+func (i *Instance) SaveSnapshot() error {
+	jsonBytes, err := json.Marshal(i.Spec)
+	if err != nil {
+		return err
+	}
+	if i.Annotations == nil {
+		i.Annotations = make(map[string]string)
+	}
+	i.Annotations[snapshotAnnotation] = string(jsonBytes)
+	return nil
+}
+
+func (i *Instance) getSnapshotedSpec() (*InstanceSpec, error) {
+	if i.Annotations != nil {
+		snapshot, ok := i.Annotations[snapshotAnnotation]
+		if ok {
+			var spec *InstanceSpec
+			err := json.Unmarshal([]byte(snapshot), &spec)
+			if err != nil {
+				return nil, err
+			}
+			return spec, nil
+		}
+	}
+	return nil, nil
+}
+
+// selectPlan returns nil if none of the plan exists, otherwise the first one in list that exists
+func selectPlan(possiblePlans []string, ov *OperatorVersion) *string {
+	for _, n := range possiblePlans {
+		if _, ok := ov.Spec.Plans[n]; ok {
+			return kudo.String(n)
+		}
+	}
+	return nil
+}
+
+// GetPlanToBeExecuted returns name of the plan that should be executed
+func (i *Instance) GetPlanToBeExecuted(ov *OperatorVersion) (*string, error) {
+	if i.GetPlanInProgress() != nil { // we're already running some plan
+		return nil, nil
+	}
+
+	// new instance, need to run deploy plan
+	if i.NoPlanEverExecuted() {
+		return kudo.String("deploy"), nil
+	}
+
+	// did the instance change so that we need to run deploy/upgrade/update plan?
+	instanceSnapshot, err := i.getSnapshotedSpec()
+	if err != nil {
```

Comment:
`getSnapshotedSpec` returns  `nil, nil` when no `annotations` or no `snapshotAnnotation` was found but you only check the `err != nil` here (which could lead to a npe below)